### PR TITLE
Emulate resource table changes and rename apk

### PR DESCRIPTION
**Pull request by amanj120**
The method _fix_resources directly edits the resource table in the template apk. To create the same effect, we have to create values.xml files in the gradle project. To rename the apk, we have to alter the build.gradle file.

### CODE REVIEW

File: platform/android/export/export.cpp [hunk 1]
```diff
@@ -795,6 +795,10 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
 		return OK;
 	}
 
+	void _fix_manifest_plaintext(const Ref<EditorExportPreset> &p_preset, Vector<uint8_t> &p_manifest, bool p_give_internet){
+
```
Comment (thanwah) [inline platform/android/export/export.cpp]: Why is this empty? Add a "TODO" to implement?

File: platform/android/export/export.cpp [hunk 2]
```diff
@@ -2232,6 +2236,9 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
 				}
 			}
 
+			new_file = new_file.replace("android_${variant.name}.apk", output_gradle_path);
+		    print_line(new_file);
```
Comment (thanwah) [inline platform/android/export/export.cpp]: Is this for debug? delete?

File: platform/android/export/export.cpp [hunk 3]
```diff
@@ -2348,26 +2355,76 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
-				return ERR_UNCONFIGURED;
-			}
-		}
-		//build project if custom build is enabled
-		String sdk_path = EDITOR_GET("export/android/custom_build_sdk_path");
-		ERR_FAIL_COND_V_MSG(sdk_path == "", ERR_UNCONFIGURED, "Android SDK path must be configured in Editor Settings at 'export/android/custom_build_sdk_path'.");
-
-		_update_custom_build_project(); //alters the build.gradle, android manifest, etc.
+	Error _copy_value_xml_files(const Ref<EditorExportPreset> &p_preset, bool p_debug){
+        //replicates the fuinctionality of _fix_resources method by renaming all project name strings.
```
Comment (thanwah) [inline platform/android/export/export.cpp]: typo: fuinctionality

File: platform/android/export/export.cpp [hunk 4]
```diff
@@ -2348,26 +2355,76 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+        String lib_file = "res://android/build/libs/debug/godot-lib.debug.aar";
+        if(!p_debug){
```
Comment (thanwah) [inline platform/android/export/export.cpp]: clang-format

File: platform/android/export/export.cpp [hunk 5]
```diff
@@ -2348,26 +2355,76 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+            data.resize(info.uncompressed_size);
+
+            //read
+            unzOpenCurrentFile(pkg);
+            unzReadCurrentFile(pkg, data.ptrw(), data.size());
+            unzCloseCurrentFile(pkg);
+
+            //write
+            if (file.begins_with("res/values") && file.ends_with(".xml")) {
+//                print_line(file);
```
Comment (thanwah) [inline platform/android/export/export.cpp]: remove such lines

File: platform/android/export/export.cpp [hunk 6]
```diff
@@ -2418,13 +2475,40 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+				return ERR_UNCONFIGURED;
+			}
+		}
+		//build project if custom build is enabled
+		String sdk_path = EDITOR_GET("export/android/custom_build_sdk_path");
+		ERR_FAIL_COND_V_MSG(sdk_path == "", ERR_UNCONFIGURED, "Android SDK path must be configured in Editor Settings at 'export/android/custom_build_sdk_path'.");
 
-        //TODO: replicate the functionality of _fix_manifest
+        String output_gradle_path = p_path.replace(ProjectSettings::get_singleton()->get_resource_path(), "")
+                                            .replace(".apk","").insert(0, "../../../../../..") + "_${variant.name}.apk";
```
Comment (thanwah) [inline platform/android/export/export.cpp]: "../../../../../.." seems a bit brittle - is there another way to get the path you need?

File: platform/android/export/export.cpp [hunk 7]
```diff
@@ -2468,16 +2552,15 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
-			src_apk = build_path.plus_file("build/outputs/apk/debug/android_debug.apk");
-		} else {
-			src_apk = build_path.plus_file("build/outputs/apk/release/android_release.apk");
-		}
 
-		if (!FileAccess::exists(src_apk)) {
-			EditorNode::get_singleton()->show_warning(TTR("No build apk generated at: ") + "\n" + src_apk);
-			return ERR_CANT_CREATE;
-		}
+//		String move_rename = "mv " + build_path + "/build/outputs/apk/debug/android_debug.apk " + p_path;
```
Comment (thanwah) [inline platform/android/export/export.cpp]: Planning to delete this code?

**Review by thanwah — CHANGES_REQUESTED** [review]

File: platform/android/export/export.cpp (hunk 6 above)
Comment (amanj120) [reply platform/android/export/export.cpp]: brittle? And I don't think so. I tried using absolute paths but I got an error saying something along the lines of "build.gradle cannot use absolute paths"

Comment (thanwah) [reply platform/android/export/export.cpp]: Can you provide an example (in a code review comment) about what an example of `ProjectSettings::get_singleton()->get_resource_path()` looks like before and after your change?

File: platform/android/export/export.cpp [hunk 8]
```diff
@@ -2418,16 +2474,40 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+				EditorNode::get_singleton()->show_warning(TTR("Trying to build from a custom built template, but no version info for it exists. Please reinstall from the 'Project' menu."));
+				return ERR_UNCONFIGURED;
+			}
+			String version = f->get_line().strip_edges();
+			if (version != VERSION_FULL_CONFIG) {
+				EditorNode::get_singleton()->show_warning(vformat(TTR("Android build version mismatch:\n   Template installed: %s\n   Godot Version: %s\nPlease reinstall Android build template from 'Project' menu."), version, VERSION_FULL_CONFIG));
+				return ERR_UNCONFIGURED;
+			}
+		}
+		//build project if custom build is enabled
```
Comment (thanwah) [inline platform/android/export/export.cpp]: Is this comment correct? I thought this code was already within the "custom build" section

File: platform/android/export/export.cpp [hunk 9]
```diff
@@ -2418,16 +2474,40 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+		_update_custom_build_project(output_gradle_path); //alters the build.gradle, android manifest, etc.
+		_copy_icon_gradle(p_preset);
+		Error copy_value_xml_err = _copy_value_xml_files(p_preset, p_debug);
+
+		String manifest_path = "res://android/build/AndroidManifest.xml";
+		Error err = _fix_manifest_plaintext(p_preset, manifest_path, p_flags & (DEBUG_FLAG_DUMB_CLIENT | DEBUG_FLAG_REMOTE_DEBUG));
 
 		APKExportData ed;
-		Error err = export_project_files(p_preset, save_gradle_project_file, &ed, save_gradle_project_so);
+		err = export_project_files(p_preset, save_gradle_project_file, &ed, save_gradle_project_so);
```
Comment (thanwah) [inline platform/android/export/export.cpp:2528]: Do you mean to overwrite "err" without checking the result from line 2507 first?

File: platform/android/export/export.cpp [hunk 10]
```diff
@@ -2348,26 +2356,74 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
 		return have_plugins_changed || first_build;
 	}
 
-	Error gradle_build(const Ref<EditorExportPreset> &p_preset, bool p_debug, const String &p_path, String &src_apk, int p_flags = 0) {
+	Error _copy_value_xml_files(const Ref<EditorExportPreset> &p_preset, bool p_debug) {
+		//replicates the functionality of _fix_resources method by renaming all project name strings.
+		String lib_file = "res://android/build/libs/debug/godot-lib.debug.aar";
```
Comment (thanwah) [inline platform/android/export/export.cpp]:
```
String lib_file;
if (p_debug) {
  lib_file = "res://android/build/libs/debug/godot-lib.debug.aar";
} else {
  lib_file = "res://android/build/libs/release/godot-lib.release.aar";
}
```

File: platform/android/export/export.cpp [hunk 11]
```diff
@@ -2348,26 +2356,74 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
-				EditorNode::get_singleton()->show_warning(vformat(TTR("Android build version mismatch:\n   Template installed: %s\n   Godot Version: %s\nPlease reinstall Android build template from 'Project' menu."), version, VERSION_FULL_CONFIG));
-				return ERR_UNCONFIGURED;
+		while (ret == UNZ_OK) {
+
+			//get filename
+			unz_file_info info;
+			char fname[16384];
+			unzGetCurrentFileInfo(pkg, &info, fname, 16384, NULL, 0, NULL, 0);
+
+			bool skip = false;
```
Comment (thanwah) [inline platform/android/export/export.cpp]: Is this variable unused?

**Review by thanwah — CHANGES_REQUESTED** [review]

File: platform/android/export/export.cpp (hunk 6 above)
Comment (amanj120) [reply platform/android/export/export.cpp]: get_resource_path returns the absolute path to the directory where the project.godot file is. p_path is the absolute path of the exported apk.

Comment (thanwah) [reply platform/android/export/export.cpp]: To make this clearer I'd suggest breaking this into a couple lines with more descriptive variable names:

```
String apk_relative_path = p_path.replace(ProjectSettings::get_singleton()->get_resource_path(), "");
String output_gradle_path = <name_here>.replace(".apk","").insert(0, "../../../../../..") + "_${variant.name}.apk";
```
Not sure if `apk_relative_path` is correct. Is `output_gradle_path` the right name for an apk path?

**Review by thanwah — APPROVED** [review]
LGTM!

**Pull request merged**